language/functional_tests/tests/testsuite/sorted_linked_list/sorted-linked-list.mvir

```diff
@@ -0,0 +1,510 @@
+
+        //make sure a node exists
+        assert(exists<Node>(copy(here)), 11);
+
+        //make sure it is not a head node (heads can be removed using remove_list)
+        assert(!exists<IsHeadNode>(copy(here)),12);
+
+        //remove it
+        Self.remove_node(move(here));
+
+        return;
+    }
+
+    public remove_node_by_list_owner(node_address: address) acquires Node
+    {
+
+        let node_ref: & Self.Node;
+        let list_owner: address;
+
+        //make sure the node exists
+        assert(exists<Node>(copy(node_address)), 13);
+
+        //make sure it is not a head node
+        assert(!exists<IsHeadNode>(copy(node_address)), 14);
+
+        //make sure the caller owns the list
+        node_ref = borrow_global<Node>(copy(node_address));
+        list_owner = *&move(node_ref).head;
+        assert(move(list_owner) == get_txn_sender(), 15);
+
+        //remove it
+        Self.remove_node(move(node_address));
+
+        return;
+    }
+
+    //private function used for removing a non-head node -- does not check permissions
+    remove_node(node_address: address) acquires Node
+    {
+        let current_node_ref: & Self.Node;
+        let next_node_address: address;
+        let next_node_mut_ref: &mut Self.Node;
+        let prev_node_address: address;
+        let prev_node_mut_ref: &mut Self.Node;
+        let temp_address: address;
+        let temp_value: u64; //TODO: make generic
+
+        //make sure the node exists
+        assert(exists<Node>(copy(node_address)),16);
+
+        //make sure it is not a head node
+        assert(!exists<IsHeadNode>(copy(node_address)),17);
+
+        //find prev and next
+        current_node_ref = borrow_global<Node>(copy(node_address));
+        next_node_address = *&copy(current_node_ref).next;
+        prev_node_address = *&move(current_node_ref).prev;
+
+        //update next
+        next_node_mut_ref = borrow_global_mut<Node>(copy(next_node_address));
+        *&mut move(next_node_mut_ref).prev = copy(prev_node_address);
+
+        //update prev
+        prev_node_mut_ref = borrow_global_mut<Node>(move(prev_node_address));
+        *&mut move(prev_node_mut_ref).next = move(next_node_address);
+
+        //destroy the current node
+        Node {temp_address,temp_address,temp_address,temp_value} = move_from<Node>(move(node_address));
+
+        return;
+    }
+
+
+}
+
+
+//! new-transaction
+//! sender: alice
+//creating a new list _@alice
+import {{sys}}.SortedLinkedList;
+main() {
+    SortedLinkedList.create_new_list();
+    return;
+}
+// check: EXECUTED
+
+//! new-transaction
+//! sender: alice
+//attempting to create another list with the same head
+import {{sys}}.SortedLinkedList;
+main() {
+    SortedLinkedList.create_new_list();
+    return;
+}
+// check: ABORTED
+
+//! new-transaction
+//! sender: bob
+//adding a new element to Alice's list _@alice -> 10@bob
+import {{sys}}.SortedLinkedList;
+main() {
+    SortedLinkedList.add_node(10, {{alice}});
```

Comment:
Might want to check that the node is in the right place here? Exposing getters in the module for this sort of thing seems useful in general.

Reply:
I added getters, but at this point the list is empty, so bob does not need to check order.

Reply:
Could you use the getters in some of the tests below? It might also be useful to add a `size` and check that the size is 1 here, the expected value elsewhere, and so on.

The thing I worry about is that these tests are testing something somewhat indirect ("do procedure calls succeed and fail where expected?") instead of something more direct about the structure/order of the list.